### PR TITLE
Implement LogManager time series filtering with TimeROI.

### PR DESCRIPTION
**Pull request by bolotovskyr**
**Description of work**
This PR implements filtering of time series properties managed by `LogManager`, according to a `TimeROI`. Time series values entirely within an ignore region of the supplied `TimeROI` will either be removed or not copied except for the last value before a use region and the first value after a use region.  Non-time series properties managed by `LogManager` will not be affected.

**To test:**
`../mantid/build/bin/APITest LogManagerTest test_removeDataOutsideTimeROI`
`../mantid/build/bin/APITest LogManagerTest test_cloneInTimeROI`

Refs #34794
<!Fixes #xxxx.--> 

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Kernel/src/PropertyManager.cpp [hunk 1]
```diff
@@ -173,6 +174,43 @@ void PropertyManager::filterByProperty(const Kernel::TimeSeriesProperty<bool> &f
...
+}
+
+/**
+ * For time series properties, remove time values outside of TimeROI regions, each defined as [roi_start,roi_stop).
+ * However, keep the values immediately before and after each ROI region, if available.
+ * @param timeROI :: a series of time regions used to determine which values to remove or to keep
+ */
+void PropertyManager::removeDataOutsideTimeROI(const Kernel::TimeROI &timeROI) {
+  for (auto prop : m_orderedProperties) {
+    auto tsp = dynamic_cast<ITimeSeriesProperty *>(prop);
```
Comment (peterfpeterson) [inline Framework/Kernel/src/PropertyManager.cpp:208]: Could be
```suggestion
    if (auto tsp = dynamic_cast<ITimeSeriesProperty *>(prop)) {
      tsp->removeDataOutsideTimeROI(timeROI);
    }
```

File: Framework/Kernel/src/PropertyManager.cpp [hunk 2]
```diff
@@ -173,6 +174,43 @@ void PropertyManager::filterByProperty(const Kernel::TimeSeriesProperty<bool> &f
...
+ */
+PropertyManager *PropertyManager::cloneInTimeROI(const Kernel::TimeROI &timeROI) {
+  PropertyManager *newMgr = new PropertyManager();
+  newMgr->m_orderedProperties.reserve(m_orderedProperties.size());
+  // We need to do a deep copy of the property pointers here
+  for (auto prop : m_orderedProperties) {
+    auto tsp = dynamic_cast<ITimeSeriesProperty *>(prop);
+    std::unique_ptr<Property> newProp;
+    if (tsp)
+      newProp = std::unique_ptr<Property>(tsp->cloneInTimeROI(timeROI));
```
Comment (peterfpeterson) [inline Framework/Kernel/src/PropertyManager.cpp:192]: I don't recall the difference between `std::unique_ptr` constructor and `std::make_unique`. Do you know if this is creating a temporary second copy?

File: Framework/Kernel/test/TimeSeriesPropertyTest.h [hunk 3]
```diff
@@ -569,11 +569,11 @@ class TimeSeriesPropertyTest : public CxxTest::TestSuite {
...
-    auto log_input = std::make_shared<TimeSeriesProperty<double>>("three_values", times, values);
-    auto log_expected = std::make_shared<TimeSeriesProperty<double>>("three_values", times_expected, values_expected);
-    auto log_result_base = std::shared_ptr<Property>(log_input->cloneInTimeROI(roi));
-    auto log_result = std::static_pointer_cast<TimeSeriesProperty<double>>(log_result_base);
-    TS_ASSERT_EQUALS(*log_result, *log_expected);
+    auto tsp_input = std::make_unique<TimeSeriesProperty<double>>("three_values", times, values);
+    auto tsp_expected = std::make_unique<TimeSeriesProperty<double>>("three_values", times_expected, values_expected);
+    auto tsp_result_base = std::shared_ptr<Property>(tsp_input->cloneInTimeROI(roi));
+    auto tsp_result = std::static_pointer_cast<TimeSeriesProperty<double>>(tsp_result_base);
+    TS_ASSERT_EQUALS(*tsp_result, *tsp_expected);
```
Comment (peterfpeterson) [inline Framework/Kernel/test/TimeSeriesPropertyTest.h:576]: Please add a check that `tsp_result` isn't equal to `tsp_input` so we know the input data isn't modified

**Review by peterfpeterson — CHANGES_REQUESTED** [review]

File: Framework/Kernel/src/PropertyManager.cpp (hunk 1 above)
Comment (bolotovskyr) [reply Framework/Kernel/src/PropertyManager.cpp]: Done.

File: Framework/Kernel/test/TimeSeriesPropertyTest.h (hunk 3 above)
Comment (bolotovskyr) [reply Framework/Kernel/test/TimeSeriesPropertyTest.h:581]: Done.

**Review by peterfpeterson — APPROVED** [review]
Looks good to me. Time to integrate into the magic 3 algorithms: `FilterEvents`, `FilterByLogValue`, and `FilterByTime`

**Pull request merged**